### PR TITLE
Fix use of Falcon language model for text generation

### LINKED ISSUE #3589: BUG: Error using Falcon for text-generation
### Issue Description

I'm trying to use SHAP with Falcon, but this issue comes out

### Minimal Reproducible Example

```python
!pip install shap
!pip install transformer
from transformers import (
    AutoModelForCausalLM,
    AutoTokenizer,
    StoppingCriteria,
    StoppingCriteriaList,
    pipeline)
import shap

MODEL_NAME = "tiiuae/falcon-7b-instruct"

model = AutoModelForCausalLM.from_pretrained(
    MODEL_NAME, trust_remote_code=True, load_in_8bit=True, device_map="auto"
)
model = model.eval()

tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)



s = ["I enjoy walking with my cute dog"]
gen_dict = dict(
    max_new_tokens=100,
    num_beams=5,
    renormalize_logits=True,
    no_repeat_ngram_size=8,
)


model.config.task_specific_params = dict()
model.config.task_specific_params["text-generation"] = gen_dict
shap_model = shap.models.TeacherForcing(model, tokenizer)

masker = shap.maskers.Text(tokenizer, mask_token="...", collapse_mask_token=True)
explainer = shap.Explainer(shap_model, tokenizer)
shap_values = explainer(s)
```

### Traceback

```shell
Setting `pad_token_id` to `eos_token_id`:11 for open-end generation.
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-7-e22bc7645a11> in <cell line: 23>()
     21 explainer = shap.Explainer(shap_model, tokenizer)
     22 # explainer = shap.Explainer(pipeline)
---> 23 shap_values = explainer(s)

9 frames
/usr/local/lib/python3.10/dist-packages/accelerate/hooks.py in new_forward(*args, **kwargs)
    163                 output = old_forward(*args, **kwargs)
    164         else:
--> 165             output = old_forward(*args, **kwargs)
    166         return module._hf_hook.post_forward(module, output)
    167 

TypeError: FalconForCausalLM.forward() got an unexpected keyword argument 'position_ids'
```

### Expected Behavior

_No response_

### Bug report checklist

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest release](https://github.com/shap/shap/releases) of shap.
- [X] I have confirmed this bug exists on the [master branch](https://github.com/shap/shap/blob/master/CONTRIBUTING.md#installing-from-the-master-branch) of shap.
- [ ] I'd be interested in making a PR to fix this bug

### Installed Versions

shap == 0.45.0

### PR DESCRIPTION
**Pull request by CloseChoice**
## Overview

Closes #3589 

Description of the changes proposed in this pull request:
- the falcon model does not need `position_ids` so we check the function signature dynamically now and only hand over the expected arguments to avoid errors
- add test with tiny falcon model (<10MB). Admittedly we can improve what we test but this test works for now. Fails on master but works on this branch.

## Checklist

- [x] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [x] Unit tests added (if fixing a bug or adding a new feature)

### CODE REVIEW

File: tests/models/test_teacher_forcing_logits.py [hunk 1]
```diff
@@ -6,6 +6,39 @@
...
+    transformers = pytest.importorskip("transformers")
+    requests = pytest.importorskip("requests")
+    name = "fxmarty/really-tiny-falcon-testing"
+    try:
+        tokenizer = transformers.AutoTokenizer.from_pretrained(name)
+        model = transformers.AutoModelForCausalLM.from_pretrained(
+            name, trust_remote_code=True, load_in_8bit=True, device_map="auto"
+        )
+    except requests.exceptions.RequestException:
+        pytest.xfail(reason="Connection error to transformers model")
```
Comment (connortann) [inline tests/models/test_teacher_forcing_logits.py:19]: Great idea, this should help our CI reliability :)

**Review by connortann — APPROVED** [review]
This looks great. Apologies for the rather slow response - I've been away for most of the past week.

I'll rename the title to something a bit more descriptive, to help the release notes.

File: tests/models/test_teacher_forcing_logits.py [hunk 2]
```diff
@@ -6,6 +6,39 @@
...
 
 
+def test_falcon():
+    transformers = pytest.importorskip("transformers")
+    requests = pytest.importorskip("requests")
+    name = "fxmarty/really-tiny-falcon-testing"
+    try:
+        tokenizer = transformers.AutoTokenizer.from_pretrained(name)
+        model = transformers.AutoModelForCausalLM.from_pretrained(
+            name, trust_remote_code=True, load_in_8bit=True, device_map="auto"
```
Comment (connortann) [inline tests/models/test_teacher_forcing_logits.py]: Ah, I noticed the tests fail on CI:

```
ImportError: Using `low_cpu_mem_usage=True` or a `device_map` requires Accelerate: `pip install accelerate`
```
  
I have a bias against adding new dev dependencies if at all possible. Can we load this model without "loc_cpu_mem_usage" : perhaps without the `load_in_8bit` kwarg?

**Review by connortann — CHANGES_REQUESTED** [review]
Pending tests passing

File: tests/models/test_teacher_forcing_logits.py (hunk 1 above)
Comment (CloseChoice) [reply tests/models/test_teacher_forcing_logits.py:19]: just stole this from the other test in the file ;) But thanks I'll take the compliment

File: tests/models/test_teacher_forcing_logits.py (hunk 2 above)
Comment (CloseChoice) [reply tests/models/test_teacher_forcing_logits.py]: done, let's see if the tests pass

File: pyproject.toml [hunk 3]
```diff
@@ -74,6 +74,8 @@ test = [
   "tensorflow<2.16;python_version<'3.12'",  # FIXME: pending keras 3.0 support, pending tf py3.12 support
   "sentencepiece",
   "opencv-python",
+  "accelerate",
+  "bitsandbytes"
```
Comment (CloseChoice) [inline pyproject.toml]: sorry, still need to remove these.

Comment (CloseChoice) [reply pyproject.toml]: done, so this is completely ready for another review round

**Review by connortann — APPROVED** [review]
LGTM!

Looks like a test failed due to a dependency install issue, which might be a transient network issue. I've retriggered that job, hopefully it will pass.

**Pull request merged**